bootstrap/src/main/java/jenkins/bootstrap/Plugin.java

```diff
@@ -0,0 +1,188 @@
+        File[] files = lib.listFiles();
+        if (files!=null) {
+            for (File jar : files) {
+                if (!jar.getName().endsWith(".jar"))
+                    continue;
+                // TODO: version check
+                jars.add(jar.toURI().toURL());
+            }
+        }
+    }
+
+    /**
+     * Gets the directory name to explode the jpi file into.
+     */
+    private File getDestDir() {
+        String name = archive.getName();
+        int idx = name.lastIndexOf('.');
+        if (idx<0)  throw new IllegalArgumentException("No extension in file name: "+archive);
+        return new File(archive.getParentFile(), name.substring(0,idx));
+    }
+
+    /**
+     * Returns true if this plugin overrides core components.
+     */
+    private boolean isCoreComponentOverride() throws IOException {
+        Manifest mf = loadManifest();
+        return "true".equals(mf.getMainAttributes().getValue("Core-Component"));
+    }
+
+    public boolean isDisabled() {
+        File disableFile = new File(archive.getPath() + ".disabled");
+        return disableFile.exists();
+    }
+
+    /**
+     * Loads jar file manifest from the archive.
+     */
+    private Manifest loadManifest() throws IOException {
+        try (JarFile jar = new JarFile(archive)) {
+            ZipEntry e = jar.getEntry("META-INF/MANIFEST.MF");
+            try (InputStream is = jar.getInputStream(e)) {
+                return new Manifest(is);
+            }
+        }
+    }
+
+    /**
+     * Unpacks {@link #archive} into {@link #destDir} with uptodate check.
+     *
+     * <p>
+     * Up-to-date check needs to be consistent with {@coe ClassicPluginStrategy} in core.
+     *
+     * <p>
+     * This simplified version does not support automatic building of classes.jar from WEB-INF/classes.
+     * Pluggable core components are not supposed to have any classes, just WEB-INF/lib files.
+     */
+    private void explode() throws IOException {
+        destDir.mkdirs();
+
+        // timestamp check
+        File explodeTime = new File(destDir,".timestamp2");
+        if(explodeTime.exists() && explodeTime.lastModified()==archive.lastModified())
+            return; // no need to expand
+
+        Project p = new Project();
+
+        if (destDir.exists()) {
+            // delete the contents so that old files won't interfere with new files
+            Delete d = new Delete();
+            d.setProject(p);
+            if (destDir.isDirectory())
+                d.setDir(destDir);
+            else
+                d.setFile(destDir);
+            d.execute();
+        }
+
+        // expand the contents
+        Expand e = new Expand();
+        e.setProject(p);
+        e.setTaskType("unzip");
+        e.setSrc(archive);
+        e.setDest(destDir);
+        e.execute();
+
+        if (new File(destDir,"WEB-INF/classes").exists()) {
+            throw new IOException("Core override plugin cannot have WEB-INF/classes: "+destDir);
```

Comment:
If the idea is to require otherwise 'empty' plugins for core override plugins, this check doesn't work. I added a trivial ItemListener to the remoting plugin, and…

	Feb 20, 2017 10:31:04 PM jenkins.bootstrap.Plugin process
	INFO: Loading core overrides from /Users/danielbeck/PR-2754-Home/plugins/remoting/WEB-INF/lib
	...
	Feb 20, 2017 10:31:09 PM org.jenkinsci.plugins.remoting.Listener onLoaded
	INFO: Loaded all items!
	Feb 20, 2017 10:31:09 PM hudson.WebAppMain$2 run
	INFO: Jenkins is fully up and running